src/java/org/apache/cassandra/net/Message.java

```diff
@@ -1285,10 +1298,18 @@ private int extractParamsSizePre40(ByteBuffer buf, int readerIndex, int readerLi
 
         private <T> int payloadSize(Message<T> message, int version)
         {
-            long payloadSize = message.payload != null && message.payload != NoPayload.noPayload
-                             ? message.verb().serializer().serializedSize(message.payload, version)
-                             : 0;
-            return Ints.checkedCast(payloadSize);
+            if (message.payload != null && message.payload != NoPayload.noPayload)
+            {
+                IVersionedAsymmetricSerializer<T,T> payloadSerializer = message.verb().serializer();
+                if (null == payloadSerializer) // per-response serializers are only required for in-jvm dtests for message filtering
+                    payloadSerializer = instance().callbacks.get(message.header.id, message.header.from).responseVerb.serializer();
+                long payloadSize = payloadSerializer.serializedSize(message.payload, version);
+                return Ints.checkedCast(payloadSize);
+            }
+            else
+            {
+                return 0;
+            }
```

Comment:
Can we also pass the payloadSerializer down to this method?

Reply:
Yes, thanks for the patch, that's how it should be.